internal/gen3fuse.go

```diff
@@ -0,0 +1,441 @@
+
+	return
+}
+
+func (fs *Gen3Fuse) GetInodeAttributes(
+	ctx context.Context,
+	op *fuseops.GetInodeAttributesOp) (err error) {
+	// Find the info for this inode.
+	info, ok := fs.inodes[op.Inode]
+	if !ok {
+		err = fuse.ENOENT
+		return
+	}
+
+	// Copy over its attributes.
+	op.Attributes = info.attributes
+
+	// Patch attributes.
+	fs.patchAttributes(&op.Attributes)
+	return
+}
+
+func (fs *Gen3Fuse) OpenDir(
+	ctx context.Context,
+	op *fuseops.OpenDirOp) (err error) {
+	// Allow opening any directory.
+	return
+}
+
+func (fs *Gen3Fuse) ReadDir(
+	ctx context.Context,
+	op *fuseops.ReadDirOp) (err error) {
+
+	// Find the info for this inode.
+	info, ok := fs.inodes[op.Inode]
+	if !ok {
+		err = fuse.ENOENT
+		return
+	}
+
+	if !info.dir {
+		err = fuse.EIO
+		return
+	}
+
+	entries := info.children
+
+	// Grab the range of interest.
+	if op.Offset > fuseops.DirOffset(len(entries)) {
+		err = fuse.EIO
+		return
+	}
+
+	entries = entries[op.Offset:]
+
+	// Resume at the specified offset into the array.
+	for _, e := range entries {
+		n := fuseutil.WriteDirent(op.Dst[op.BytesRead:], e)
+		if n == 0 {
+			break
+		}
+
+		op.BytesRead += n
+	}
+
+	return
+}
+
+func (fs *Gen3Fuse) OpenFile(
+	ctx context.Context,
+	op *fuseops.OpenFileOp) (err error) {
+
+	FuseLog("inside OpenFile")
+
+	info, ok := fs.inodes[op.Inode]
+	if !ok {
+		err = fuse.ENOENT
+		return
+	}
+
+	if info.presignedUrl == "" || len(info.presignedUrl) < 3 {
+		presignedUrl, err := fs.GetPresignedURL(info.DID)
+		if err != nil {
+			return err
+		} else {
+			info.presignedUrl = presignedUrl
+			var fileBody string = FetchContentsAtURL(info.presignedUrl)
+			var fileSize = uint64(len(fileBody))
+			info.attributes.Size = fileSize
+			info.fileBody = fileBody
+			fs.inodes[op.Inode] = info
+		}
+	}
+
+	fs.inodes[op.Inode] = info
+
+	return
+}
+
+func (fs *Gen3Fuse) ReadFile(
+	ctx context.Context,
+	op *fuseops.ReadFileOp) (err error) {
+
+	info, ok := fs.inodes[op.Inode]
+	if !ok {
+		err = fuse.ENOENT
+		return
+	}
+
+	reader := strings.NewReader(info.fileBody)
+
+	op.BytesRead, err = reader.ReadAt(op.Dst, op.Offset)
+
+	// Clear out the file contents; we don't want all these files stored in memory
+	info.fileBody = ""
+	fs.inodes[op.Inode] = info
+
+	// Special case: FUSE doesn't expect us to return io.EOF.
+	if err == io.EOF {
+		err = nil
+	}
+
+	return
+}
+
+type presignedURLResponse struct {
+	Url string
+}
+
+func (fs *Gen3Fuse) GetPresignedURL(DID string) (presignedUrl string, err error) {
+	resp, err := fs.FetchURLResponseFromFence(DID)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode == 200 {
+		return fs.URLFromSuccessResponseFromFence(resp), nil
+	} else if resp.StatusCode == 401 {
+		// get a new access token, try again just one more time
+		FuseLog("Got 401, retrying...")
+		fs.accessToken = GetAccessToken(fs.gen3FuseConfig)
+		respRetry, err := fs.FetchURLResponseFromFence(DID)
+		if err != nil {
+			return "", err
```

Comment:
When fence responses with an 401 error, it will return a html page showing the error. We should handle this error message better by providing some useful messages explaining what had happened rather then dumping the whole html page into error log.

Reply:
I agree! But Fence returns this error page instead of a clean JSON error, so if I was to parse the html page to obtain the exact error message, the FUSE code would be highly coupled to their UI design. For debugging purposes, I think it's helpful to have the full error :(